### PR TITLE
189 Add HoldNote objects

### PR DESCRIPTION
**Pull request by Xenocidel**

### CODE REVIEW

File: S2VX.Game/Story/Note/EditorHoldNote.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+            switch (NumHitSounds) {
+                case 2:
+                    if (time >= HitTime) {
+                        --NumHitSounds;
+                        Hit.Play();
+                    }
+                    break;
+                case 1:
+                    if (time >= EndTime) {
```
Comment (Xenocidel) [inline S2VX.Game/Story/Note/EditorHoldNote.cs:54]: I realize I can collapse this down into one

**Review by Naranja-Sagged — CHANGES_REQUESTED** [review]
HoldNotes get stuck on the left side of the notesTimeline when current time is past the HoldNote's endTime
image

**Review by Naranja-Sagged — CHANGES_REQUESTED** [review]
Pressing another input and releasing that same input too early will count as a miss even if user is still holding down the original input they initially pressed on the HoldNote. I think having at least one input held down should prevent a miss, like how you can hold and switch inputs during a slider in osu!.

File: S2VX.Game/Editor/Containers/NotesTimeline.cs [hunk 2]
```diff
@@ -259,21 +259,43 @@ public class NotesTimeline : CompositeDrawable {
...
-                        Height = TimelineNoteHeight,
-                        X = (float)relativePosition,
-                        Y = 0.2f,
-                        Anchor = Anchor.TopLeft,
-                    };
+                RelativeBox visibleNote = null;
+                var relativePosition = (note.HitTime - lowerBound) / sectionDuration;
+
+                if (note is EditorHoldNote holdNote) {
+                    if (lowerBound <= holdNote.HitTime || holdNote.EndTime <= upperBound) {
```
Comment (Naranja-Sagged) [inline S2VX.Game/Editor/Containers/NotesTimeline.cs:271]: The HoldNote getting stuck on the left side of NotesTimeline probably has something to do with the logic here. I tried changing the || to && but the HoldNote disappears too quickly Lol

**Comment by maxrchung** [comment]
Pausing my review until Alex's issues are addressed. I don't think I have anything else to add though, code structure looks fine.

File: S2VX.Game/Editor/Containers/NotesTimeline.cs (hunk 2 above)
Comment (Xenocidel) [reply S2VX.Game/Editor/Containers/NotesTimeline.cs:271]: Fixed

**Comment by Xenocidel** [comment]
> Pressing another input and releasing that same input too early will count as a miss even if user is still holding down the original input they initially pressed on the HoldNote. I think having at least one input held down should prevent a miss, like how you can hold and switch inputs during a slider in osu!.

I tried changing IsBeingHeld to be NumInputsHeld but KeyDownEvent gets called multiple times if you hold down a key (key repeats as if it were a text editor). Need to figure that out.

**Comment by maxrchung** [comment]
> Pressing another input and releasing that same input too early will count as a miss even if user is still holding down the original input they initially pressed on the HoldNote. I think having at least one input held down should prevent a miss, like how you can hold and switch inputs during a slider in osu!.

Aaron and I discussed switching inputs in the middle of a hold/slider. I think the key you start a hold/slider with should be the same key you release with. This way also seems a bit easier to implement compared to allowing switching inputs in the middle of the hold/slider.

**Review by maxrchung — APPROVED** [review]

**Review by Naranja-Sagged — APPROVED** [review]

**Pull request merged**